File: web/src/lib/import/duplicate-detector.ts
import { fingerprintRepo } from '@/lib/repos';
import type { RawTransaction } from '@/lib/types';

/**
 * Mark transactions as duplicates if their fingerprint already exists for
 * this account, or appears more than once within the same batch.
 *
 * One bulk DB lookup instead of one-per-row — for a 500-row import that's
 * the difference between 500 round-trips and 1.
 */
export async function checkDuplicates(
  transactions: RawTransaction[],
  accountId: number,
): Promise<RawTransaction[]> {
  // De-duplicate the fingerprint list before the IN(...) lookup —
  // shrinks parameter count and query size for large imports without
  // changing behaviour (a fingerprint either exists for the account or
  // it doesn't).
  const fingerprints = [...new Set(transactions.map(t => t.fingerprint))];
  const existingInDb = await fingerprintRepo.findExisting(fingerprints, accountId);

  const seenInBatch = new Set<string>();
  for (const txn of transactions) {
    if (existingInDb.has(txn.fingerprint) || seenInBatch.has(txn.fingerprint)) {
      txn.isDuplicate = true;
    } else {
      seenInBatch.add(txn.fingerprint);
    }
  }

  return transactions;
}

File: finadviser/importing/duplicate_detector.py
"""Duplicate transaction detection."""

from __future__ import annotations

import sqlite3

from finadviser.db.models import RawTransaction
from finadviser.db.repositories import FingerprintRepo


class DuplicateDetector:
    """Marks transactions as duplicates if their fingerprint already exists for the account."""

    def __init__(self, conn: sqlite3.Connection) -> None:
        self.fp_repo = FingerprintRepo(conn)

    def check(self, transactions: list[RawTransaction], account_id: int) -> list[RawTransaction]:
        """Check each transaction for duplicates and set is_duplicate flag.

        Also detects duplicates within the batch itself.
        """
        seen_in_batch: set[str] = set()

        for txn in transactions:
            if self.fp_repo.exists(txn.fingerprint, account_id):
                txn.is_duplicate = True
            elif txn.fingerprint in seen_in_batch:
                txn.is_duplicate = True
            else:
                seen_in_batch.add(txn.fingerprint)

        return transactions
